### ISSUE TITLE
Instrument AccordionBox with A11y better

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/sun#444]
The instrumentation of a11y in AccordionBox is strange. It adds instrumentation to the title bar, so `expandedTitleBar` and `collapsedTitleBar` are the nodes that receive focus. One issue is that the title bar can be invisible, and adding a11y related input listeners is making the title bar look pickable.  One workaround could be to make the title bars non-pickable depending on options `showTitleWhenExpanded` and `titleBarExpandCollapse`.

But we should also not have the a11y instrumentation be on the title. And from https://github.com/phetsims/tasks/issues/980 we should generally reconsider how a11y should be done for AccordionBox. It also shouldn't be done until #359 is complete.

The A11y team won't get to this for at least three months, so I propose we remove a11y from AccordionBox to fix the immediate issues instead of adding workarounds to AccordionBox.

**Comment by pixelzoom** [comment]
Over in https://github.com/phetsims/balancing-chemical-equations/issues/125, we decided to remove a11y from AccordionBox since it's causing problems in production code.

**Comment by zepumph** [comment]
We are now ready to work on AccordionBox more generally again because Ratio and Proportion needs to use this. I will take a first pass and refamiliarize myself with AccordionBox.md as it pertains to the implementation  in AccordionBox.js, and will likely have questions for @jessegreenberg that we can work on together tomorrow.

**Comment by zepumph** [comment]
I read through all of the BCE issue, and found that a good summary of the a11y issues/changes that should be done is in https://github.com/phetsims/balancing-chemical-equations/issues/125#issuecomment-453650245:

@pixelzoom said:
>I misunderstood.  @jessegreenberg and I discussed on Slack.  AccordionBox is in a non-final state, see https://github.com/phetsims/sun/issues/444.   My opinion is that (1) keyboard focus should skip the title bar and should be on the expand/collapse button in all cases, and (2) we should consider removing the `titleBarExpandCollapse` option altogether.  But that will require design discussion. So in the meantime, we should probably yank a11y from AccordionBox in the 1.2 release branch (and master?).

RE:
> keyboard focus should skip the title bar and should be on the expand/collapse button in all cases

I think that is a good place to start. My first thought is that we don't want to be toggling visibility on focusable elements. Perhaps the ExpandCollapseButton can be the focusable button with navigation. I will take a first pass!

**Comment by zepumph** [comment]
- [ ] I will hold off on the `aria-controls` attribute because it is another "association" style attribute, like aria-describedby, and each added one is about 150 lines until https://github.com/phetsims/scenery/issues/952 can be worked on. I don't want to add that immedtiately.

**Comment by zepumph** [comment]
- [ ] In the design html in AccordionBox.md, it was said that when the box is not expanded, that the content be `aria-hidden=true`, but actually this content is just getting removed from the PDOM. Is that ok? My understanding is that this will yield the exact same effect when using AT, but it would be good to double check.

**Comment by zepumph** [comment]
- [ ] I see a design question: "Accordion Title Hidden?". I am wondering what the label of the button will be if the title get's hidden. Do you want me to add some sort of default value like "Expand Me!"?

**Comment by zepumph** [comment]
I think I am ready for commit. The changes were not as intrusive as I thought they might become. And I say this recognizing that we don't want to skirt around common code implementations anymore (as decided in https://github.com/phetsims/tasks/issues/980). 

I think that much of the issue that was run into with the implementation removed in https://github.com/phetsims/sun/commit/614e324bed1a8a60d491406a2fd73be76b0e808c was that we were trying to implement custom alternative input directly into AccordionBox. The implementation committed above instead just used the ExpandCollapseButton as the base interaction, and AccordionBox basically just wrapped that. 


Next I will look into focus control.

**Comment by zepumph** [comment]
FocusHighlight was added above. I only based it on the expandedTitleBar dimensions. I could have swapped out the base titleBar depending on if the box was expanded or collapsed, but it seemed more complicated than was needed.  Let me know if you feel differently.

I'm pretty sure we want the whole title bar to highlight even if there is no title content, like this:

![image](https://user-images.githubusercontent.com/6856943/87357055-b3cd9f00-c50f-11ea-9c72-f299fe3899b2.png)


Let me know if that is for some reason not right.


I am done with this implementation pre-review. I would like @terracoda to have a look at (or delegate to the right person) the above three check boxes in addition to the above thoughts in this comment. I feel like these are design questions. Will you please comment on them, while having a listen on master at the current AccordionBox description implementation. Thanks!

**Comment by terracoda** [comment]
In https://github.com/phetsims/sun/issues/444#issuecomment-657800821, @zepumph asked
>  "Accordion Title Hidden?". I am wondering what the label of the button will be if the title get's hidden. Do you want me to add some sort of default value like "Expand Me!"?

The non-visual experience always has accessible heading content, and then the code communicates "expanded" or "collapsed." If visually, the design has no heading, we will still need to create a relevant heading for the non-visual experience.

**Comment by terracoda** [comment]
@zepumph, where do I see it in action? I'll check RaP.

**Comment by zepumph** [comment]
>we will still need to create a relevant heading for the non-visual experience.

Hopefully I just implemented that! See the second screen of Ratio and Proportion.

**Comment by terracoda** [comment]
It sounds and looks good to me in RaP.

**Comment by zepumph** [comment]
@terracoda is your review complete?

**Comment by terracoda** [comment]
It seems to work fine.

**Comment by terracoda** [comment]
Nice job.

**Comment by terracoda** [comment]
There is one thing I am not sure about...
whether the heading should be around the button or the button around the heading?
Is that explained in the MD file?

**Comment by terracoda** [comment]
The MD file says:
> The typical structure is a button nested within the parent titling element, likely a heading.

And that's what you have coded.
![Screen Shot 2020-07-14 at 6 28 18 AM](https://user-images.githubusercontent.com/1902116/87406453-4e26f480-c59b-11ea-9165-0a82df383118.png)

I think the implementation of the pattern is in order.

**Comment by zepumph** [comment]
Above. @jessegreenberg and I made an improvement to the common code focusHighlight strategy used in AccordionBox. We made it more graceful on startup, and so that it auto updates. I also asked @jessegreenberg to take a look at the changes made in AccordionBox to see how they balance simplicity, feature encapsulation, and also not "tip toeing" around code already implemented in AccordionBox (RE https://github.com/phetsims/tasks/issues/980). I will assign him when review is ready.

@terracoda it may be easiest if we had a short meeting to discuss the above design checkboxes. Do you think that is how we should proceed?

**Comment by jessegreenberg** [comment]
The assertion added in https://github.com/phetsims/scenery/commit/0a7037cb5eaaa1bd31f574f37bd9e79b0060cc56 introduced an error in GrabDragInteractionTests. Not sure yet if it is an issue with the assertion or GrabDragInteraction yet, but I am out of time for now. So that tests are passing and commit hooks can run I am going to remove the assertion for now and come back to this tomorrow.

**Comment by zepumph** [comment]
Thanks for looking into things! If you get stuck we should talk next week at our pairing time.

**Comment by jessegreenberg** [comment]
Thanks @zepumph - I am pretty stumped, and I suspect that there may be something tricky going on with QUnit or something. Here is an example stack trace that I think is causing problems and is really confusing me:
![image](https://user-images.githubusercontent.com/6396244/87580685-ddb8ca00-c6a5-11ea-9751-7cf65677147d.png)

- GrabDragInteractionTests - line 41: we create a new GrabDragInteraction for a rectangle. The node passed in has defined bounds.
- FocusHighlightFromNode - line 115: Link the boundsListener to FocusHighlightFromNode.observedBoundsProperty. The FocusHighlightFromNode.observedBoundsProperty.value has finite bounds.
![image](https://user-images.githubusercontent.com/6396244/87581845-96cbd400-c6a7-11ea-98a7-595ba9d8e34e.png)
- FocusFromHighlightNode - line 97: The bounds passed to the callback are NOT finite. So we return early and never set the shape.

Even weirder, if I query `this.observedBoundsProperty` in the Chrome dev tools, it updates the value and the bounds become defined for the whole stack.

**Comment by jessegreenberg** [comment]
I added the assertion back in, and a workaround for https://github.com/phetsims/axon/issues/309. Tests are passing again. But keeping my assignment to review the changes listed in https://github.com/phetsims/sun/issues/444#issuecomment-658386391.

**Comment by terracoda** [comment]
@zepumph, I am not sure what you are saying about checkboxes. There are no checkboxes in RaP.

As for the AccordonBox in RaP with the 2 number spinners, everything is sounding good to me.

I think the help text for the number spinners should come before the hand number spinners, so I will create a separate issue for that in the RaP repo.
I ran what I think is the generated HTML through the W3C validator and I only got 3 warnings, no errors:
![Screen Shot 2020-07-16 at 10 37 56 AM](https://user-images.githubusercontent.com/1902116/87674461-9c312900-c750-11ea-9b5c-37f0e9358058.png)

So, I think the implementation of AccordionBox is good.
Do you think AccordionBox.md needs any further clarification? Maybe around blank heading?

**Comment by terracoda** [comment]
@zepumph, I read through the AccordionBox.md file and made a tweak to make it clear a an accessible name is needed for the Accordion title whether it is visible or not.

I have not tested on iOS. I need a dev version if you need me to test on iOS.

**Comment by terracoda** [comment]
I unassigned. Please reassign if any further testing is needed.

**Comment by zepumph** [comment]
@terracoda there are three checkboxes in github comments above, the first one is https://github.com/phetsims/sun/issues/444#issuecomment-657785511. Can you speak to those three questions:

- [ ] I will hold off on the `aria-controls` attribute because it is another "association" style attribute, like aria-describedby, and each added one is about 150 lines until https://github.com/phetsims/scenery/issues/952 can be worked on. I don't want to add that immedtiately.
UPDATE: How important is this feature for AccordionBox? Does it need to be added now?

- [ ] In the design html in AccordionBox.md, it was said that when the box is not expanded, that the content be `aria-hidden=true`, but actually this content is just getting removed from the PDOM. Is that ok? My understanding is that this will yield the exact same effect when using AT, but it would be good to double check. 
UPDATE:
>but actually this content is just getting removed from the PDOM.

This is not true, actually the content that is not visible has the `hidden` attribute added to it. Still though, the same confirmation applies


- [ ] I see a design question: "Accordion Title Hidden?". I am wondering what the label of the button will be if the title get's hidden. Do you want me to add some sort of default value like "Expand Me!"?

I don't think that you need to test on iOS, but here is a dev version just in case you want to: https://phet-dev.colorado.edu/html/ratio-and-proportion/1.0.0-dev.37/phet/ratio-and-proportion_all_phet.html

**Comment by terracoda** [comment]
@zepumph, we don't need to implement `arai-controls` now, and maybe never. I haven't recently checked on how well it is supported by AT, but last time I checked, it was "[poop](https://heydonworks.com/article/aria-controls-is-poop/)" (2016, heydonworks.com). AND our content panels always come right after the accordion title panel (i.e., accordion header), so we don't need `aria-controls` even if it works.

- checking first question as done.

Re: The Accordion Title has to be created for the PDOM. There is a visual design option to have it visually hidden, but the title will not be hidden in the PDOM.

```
<h3><button aria-expanded="true/false">My Visually Hidden Title</button></h3>
<!-- accordion panel content -->
```
I made a few tweaks to the AccordionBox.md file, including removing `aria-controls` from the code samples. 
See 5c1877e05088ccb68945ac7c7cac991e6ecaa43c .

You could add a default string for the title. "Expand Me" works, and hopefully designers won't rely on that. It really should be relevant to the content of the sim.

- checking next question as done.

**Comment by terracoda** [comment]
Re second question: "but actually this content is just getting removed from the PDOM. Is that ok?"

As long as the content is hidden to all learners and communicated as hidden, the method of the hiding likely does not matter.

**Comment by terracoda** [comment]
I unassigned myself again. Hopefully, I got everything this time :-)

**Comment by zepumph** [comment]
I removed the aria-controls from the AccordionBox.md documentation, stating that we don't need it right now. 


>You could add a default string for the title. "Expand Me" works, and hopefully designers won't rely on that. It really should be relevant to the content of the sim.

I think that providing a title seems like the best way to move forward. I mentioned this in https://github.com/phetsims/joist/issues/511 to see if we could guarantee that all AccordionBoxes follow this.

Thanks for the review! 

Next I will look at @jessegreenberg's comments above.

**Comment by zepumph** [comment]
This is not really ready for review. It is buggy and needs fixing.

**Comment by zepumph** [comment]
I looked at @jessegreenberg's comments from three weeks back, and I think things will be sorted out in https://github.com/phetsims/axon/issues/309. @jessegreenberg is there anything else for this issue?

**Comment by zepumph** [comment]
Instead of using the `setAccessibleName` override pattern used in comboBox, Now that https://github.com/phetsims/scenery/issues/1070 was fixed, we should use a behavior function for AccordionBox.

**Comment by zepumph** [comment]
The above commit should not have changed any part of the rendered PDOM, but is a much better pattern. @jessegreenberg please include it with your review of this issue.

**Comment by jessegreenberg** [comment]
I reviewed the PDOM code in AccordionBox, it looks really nice. And the rendered content matches what is requested in AccordionBox.md. https://github.com/phetsims/sun/commit/ef901e4e0999b9a2322b886f5ce1190716f17ad5 is great and just using accessibleName is really nice at usage sites, I hope we can continue to use it more.

Closing this issue.
